hardhat/contracts/LilypadPow.sol

```diff
@@ -0,0 +1,153 @@
+    address[] public miners;
+
+    mapping(address => Challenge) public lastChallenges;
+    uint256 public validProofs;
+    uint256 public startTime;
+
+    uint256 public window_start;
+    uint256 public window_end;
+    /**
+     * Init
+     */
+
+    // https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable
+    function initialize() public initializer {}
+
+    function getMiners() external view returns (address[] memory) {
+        return miners;
+    }
+
+    // generateChallenge gen a byte32 value as challenge value, Sc store this one for verify
+    function generateChallenge(string calldata nodeId) external {
+        checkTimeWindow();
+
+        bytes32 challenge = keccak256(
+            abi.encodePacked(block.timestamp, window_start, msg.sender, nodeId)
+        );
+
+        uint256 difficulty = calculate_difficulty();
+        lastChallenges[msg.sender] = Challenge(
+            challenge,
+            difficulty,
+            nodeId,
+            block.timestamp
+        );
+        emit GenerateChallenge(challenge, difficulty);
+    }
+
+    function calculate_difficulty() public view returns (uint256) {
+        uint256 percentChange = 90 + (block.prevrandao % 21);
+        return (TARGET_DIFFICULTY * percentChange) / 100;
+    }
+
+    // submitWork miner submint a nonce value, sc check the difficulty and emit a valid pow event when success
+    function submitWork(uint256 nonce, string calldata nodeId) external {
+        checkTimeWindow();
+
+        Challenge memory lastChallenge = lastChallenges[msg.sender];
+        bytes32 challenge = keccak256(
+            abi.encodePacked(
+                lastChallenge.timestamp,
+                window_start,
+                msg.sender,
+                nodeId
+            )
+        );
+
+        require(
+            lastChallenge.challenge == challenge,
+            "Work submit not compatable with challenge"
+        );
+
+        bytes32 hash = keccak256(abi.encodePacked(challenge, nonce));
+        require(
+            uint256(hash) < lastChallenge.difficulty,
+            "Work does not meet difficulty target"
+        );
+
+        validProofs++;
+
+        if (minerSubmissionCount[msg.sender] == 0) {
+            //first submit, append to miners
+            miners.push(msg.sender);
+        }
+
+        minerSubmissionCount[msg.sender]++; //increase miner's valid proofs
+        POWSubmission[] storage posSubmissions = powSubmissions[msg.sender];
+        posSubmissions.push(
+            POWSubmission(
+                msg.sender,
+                nodeId,
+                nonce,
+                lastChallenge.timestamp,
+                block.timestamp,
+                lastChallenge.challenge,
+                lastChallenge.difficulty
+            )
+        );
+
+        //clean last challenge to submit the same proof
+        lastChallenges[msg.sender] = Challenge(0, 0, "", 0);
+        emit ValidPOWSubmitted(
```

Comment:
@hunjixin I think we need to emit the hashrate here too, because we're going to base our rewards calculations off of the data that is emitted here

Reply:
This can be calculated offline, as there are some advanced computations(a quite big high pow) that I am not sure can be done in Solidity. It might result in a large amount of gas fees.